apps/src/lab2/views/components/Instructions/NavigationButton.tsx

```diff
@@ -0,0 +1,185 @@
+import {Button} from '@code-dot-org/component-library/button';
+import {ComponentSizeXSToL} from '@code-dot-org/component-library/common/types';
+import React from 'react';
+
+import {sendSubmitReport} from '@cdo/apps/code-studio/progressRedux';
+import {
+  getCurrentLevel,
+  nextLevelId,
+} from '@cdo/apps/code-studio/progressReduxSelectors';
+import {queryParams} from '@cdo/apps/code-studio/utils';
+import continueOrFinishLesson from '@cdo/apps/lab2/progress/continueOrFinishLesson';
+import {isPredictResponseSubmitted} from '@cdo/apps/lab2/redux/predictLevelRedux';
+import {LevelProperties} from '@cdo/apps/lab2/types';
+import {commonI18n} from '@cdo/apps/types/locale';
+import {logUserLevelInteraction} from '@cdo/apps/userLevelInteractionsLogger/userLevelInteractionsApi';
+import {useAppDispatch, useAppSelector} from '@cdo/apps/util/reduxHooks';
+import {
+  LevelStatus,
+  UserLevelInteractions,
+} from '@cdo/generated-scripts/sharedConstants';
+
+import {DialogType, useDialogControl} from '../../dialogs';
+
+interface NavigationButtonProps {
+  levelProperties: LevelProperties;
+  hasRun: boolean;
+  hasEdited: boolean;
+  className?: string;
+  size?: ComponentSizeXSToL;
+}
+
+const NavigationButton: React.FC<NavigationButtonProps> = ({
+  levelProperties,
+  hasRun,
+  hasEdited,
+  className,
+  size,
+}) => {
+  if (levelProperties.submittable) {
+    return (
+      <SubmitButton
+        levelId={levelProperties.id}
+        appName={levelProperties.appName}
+        hasRun={hasRun}
+        hasEdited={hasEdited}
+        className={className}
+      />
+    );
+  }
+
+  return (
+    <ContinueButton
+      className={className}
+      size={size}
+      isPredictLevel={levelProperties.predictSettings?.isPredictLevel}
+    />
+  );
+};
+
+interface ContinueButtonProps {
+  className?: string;
+  size?: ComponentSizeXSToL;
+  isPredictLevel?: boolean;
+}
+
+/**
+ * Displays the "Continue" or "Finish" button that advances to the next level or finishes the progression.
+ */
+const ContinueButton: React.FC<ContinueButtonProps> = ({
+  className,
+  size,
+  isPredictLevel,
+}) => {
+  const dispatch = useAppDispatch();
+  const hasNextLevel = useAppSelector(
+    state => nextLevelId(state) !== undefined
+  );
+  const hasSubmittedPredictResponse = useAppSelector(
+    isPredictResponseSubmitted
+  );
+  const {hasConditions, satisfied} = useAppSelector(
+    state => state.lab.validationState
+  );
```

Comment:
philosophical question: `InstructionsPanel` is explicitly not using redux, and we pass a bunch of values from redux into it so it is independent. Should we be passing these redux values in instead of fetching them here so `InstructionsPanel` avoids getting things from redux? I think a couple predict components are also breaking this paradigm, so it could be we decide it's ok to have redux values in dependencies of `InstructionsPanel`. What do you think?

Reply:
I think [this usage](https://github.com/code-dot-org/code-dot-org/pull/66289) of the component expects no redux, right @bencodeorg?

Reply:
Good question, and yeah should have called this out in the description 🙂 I did verify and this _happens_ to work by virtue of `canShowNextButton` being undefined (so `NavigationButton` button is never rendered). But yes this was on my mind too - I was thinking we could tackle this as part of the instructions unification work?

One idea was, if we do end up splitting out the various components of the instructions (i.e. main text, validations, predict question, exemplar player, navigation button, etc), then the levelbuilder `initializeCodeMirror` file can just render the main text portion directly (presumably it would only require a single `markdownText` prop).